File: implementations/spring-boot-starter/src/test/kotlin/io/github/mtempty/ashurbanipal/ReferencedByIntegrationTest.kt
package io.github.mtempty.ashurbanipal

import tools.jackson.databind.JsonNode
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.springframework.http.HttpStatus
import org.springframework.web.client.HttpClientErrorException
import java.net.URI
import java.net.URLEncoder
import java.sql.DriverManager
import java.sql.Statement

/**
 * `GET /api/tables/referenced-by` (`spec/protocol.md` §5.9) against the
 * devcontainer's live Postgres with `conformance/seed/seed.sql`:
 * `public.users` is referenced by several tables incl. cross-schema
 * `warehouse.shipment_events`; `inventory_locations` by exactly one composite
 * FK; `feature_flags` by nothing.
 */
class ReferencedByIntegrationTest : AshurbanipalHttpTestBase() {

    private fun referencedBy(table: String): JsonNode =
        getJson("/api/tables/referenced-by?table=$table")["referenced_by"]

    private fun entriesFor(list: JsonNode, table: String): List<JsonNode> =
        (list as Iterable<JsonNode>).filter { it["table"].asText() == table }

    private fun onlyPair(entry: JsonNode): Pair<String, String> {
        assertEquals(1, entry["columns"].size(), "expected a single column pair in $entry")
        val c = entry["columns"].first()
        return c["from"].asText() to c["to"].asText()
    }

    private fun expectBadRequest(pathSuffix: String) {
        val ex = assertThrows(HttpClientErrorException::class.java) {
            http.getForObject(URI(url("/api/tables/referenced-by") + pathSuffix), String::class.java)
        }
        assertEquals(HttpStatus.BAD_REQUEST, ex.statusCode)
        // §5.9: the protocol-version header rides the error response too.
        assertEquals("1", ex.responseHeaders?.getFirst("x-ashurbanipal-protocol"))
    }

    @Test
    fun `lists incoming FK constraints with explicit column pairs`() {
        val list = referencedBy("users")

        val orders = entriesFor(list, "orders")
        assertEquals(1, orders.size)
        assertEquals("user_id" to "id", onlyPair(orders.first()))
        assertTrue(orders.first()["constraint"].asText().isNotEmpty())
        assertFalse(orders.first().has("schema"), "same-schema referrer must omit schema")

        // support_tickets has two separate FKs into users → two entries.
        val tickets = entriesFor(list, "support_tickets")
        assertEquals(2, tickets.size)
        assertEquals(
            setOf("user_id" to "id", "assigned_admin_id" to "id"),
            tickets.map { onlyPair(it) }.toSet(),
        )
        tickets.forEach { assertFalse(it.has("schema")) }

        // (table, constraint) is unique within one response.
        val keys = (list as Iterable<JsonNode>).map { it["table"].asText() to it["constraint"].asText() }
        assertEquals(keys.size, keys.toSet().size, "duplicate (table, constraint) in $list")
    }

    @Test
    fun `composite foreign keys are included with every column pair`() {
        val list = referencedBy("inventory_locations")
        assertEquals(1, list.size(), "inventory_locations has exactly one referrer")
        val entry = list.first()
        assertEquals("inventory_counts", entry["table"].asText())
        val pairs = (entry["columns"] as Iterable<JsonNode>).map { it["from"].asText() to it["to"].asText() }
        assertEquals(
            listOf("warehouse_code" to "warehouse_code", "bin_code" to "bin_code"),
            pairs,
            "column pairs must be in the constraint's own column order",
        )
    }

    @Test
    fun `a table nothing references returns an empty list`() {
        assertEquals(0, referencedBy("feature_flags").size())
    }

    @Test
    fun `cross-schema referrers carry a schema field`() {
        val se = entriesFor(referencedBy("users"), "shipment_events")
        assertEquals(1, se.size)
        assertEquals("warehouse", se.first()["schema"].asText())
        assertEquals("handled_by_user_id" to "id", onlyPair(se.first()))
    }

    @Test
    fun `explicit schema=public matches the implicit default`() {
        assertEquals(
            getJson("/api/tables/referenced-by?table=users"),
            getJson("/api/tables/referenced-by?schema=public&table=users"),
        )
    }

    @Test
    fun `unknown or malicious table values are rejected cleanly`() {
        for (bad in listOf("", "no_such_table", "users\"; drop table users; --", "users' OR '1'='1")) {
            expectBadRequest("?table=" + URLEncoder.encode(bad, "UTF-8"))
        }
        expectBadRequest("") // no table param at all
    }

    @Test
    fun `unknown schema and unknown source are rejected cleanly`() {
        expectBadRequest("?schema=no_such_schema&table=users")
        expectBadRequest("?source=no_such_source&table=users")
    }

    @Test
    fun `the 200 response carries the protocol version header`() {
        val response = http.getForEntity(url("/api/tables/referenced-by?table=users"), String::class.java)
        assertEquals("1", response.headers.getFirst("x-ashurbanipal-protocol"))
    }

    /**
     * Regression tests: Postgres copies an inherited FK constraint onto
     * every partition, and `list_tables`' `information_schema` gate is
     * broader than the `relkind = 'r'` filter §5.2 (and so §5.9's own
     * table-gate) is meant to enforce. Each test uses its own schema —
     * disjoint names, since JUnit may run these methods in any order and a
     * shared schema stays open across the whole test class otherwise.
     */
    private fun withAdminConnection(databaseUrl: String, block: (Statement) -> Unit) {
        val uri = URI(databaseUrl)
        val (user, password) = (uri.userInfo ?: ":").split(":", limit = 2).let { it[0] to it.getOrElse(1) { "" } }
        val jdbcUrl = "jdbc:postgresql://${uri.host}:${uri.port}${uri.path}"
        DriverManager.getConnection(jdbcUrl, user, password).use { conn ->
            conn.createStatement().use(block)
        }
    }

    private fun setupPartitionedSchema(schema: String) {
        val databaseUrl = System.getenv("DATABASE_URL")
            ?: error("DATABASE_URL must be set (the devcontainer sets it automatically)")
        withAdminConnection(databaseUrl) { st ->
            st.execute("drop schema if exists $schema cascade")
            st.execute("create schema $schema")
            st.execute("create table $schema.users (id int primary key)")
            st.execute(
                "create table $schema.events (" +
                    "    id bigint not null, " +
                    "    user_id int references $schema.users(id)" +
                    ") partition by range (id)",
            )
            st.execute("create table $schema.events_p1 partition of $schema.events for values from (1) to (1000)")
            st.execute("create table $schema.events_p2 partition of $schema.events for values from (1000) to (2000)")
        }
    }

    private fun dropSchema(schema: String) {
        val databaseUrl = System.getenv("DATABASE_URL") ?: return
        withAdminConnection(databaseUrl) { st -> st.execute("drop schema if exists $schema cascade") }
    }

    @Test
    fun `reports one entry for a partitioned referrer, not one per partition`() {
        val schema = "ashb_test_spring_referenced_by_partitioning_dup"
        setupPartitionedSchema(schema)
        try {
            val list = getJson("/api/tables/referenced-by?schema=$schema&table=users")["referenced_by"]
            val events = (list as Iterable<JsonNode>).filter { it["table"].asText().startsWith("events") }
            assertEquals(
                1,
                events.size,
                "a partitioned referrer's inherited FK copies must collapse to one entry, got $events",
            )
            assertEquals("events", events.first()["table"].asText())
        } finally {
            dropSchema(schema)
        }
    }

    @Test
    fun `rejects a partitioned table as target the same as any other unlisted table`() {
        val schema = "ashb_test_spring_referenced_by_partitioning_gate"
        setupPartitionedSchema(schema)
        try {
            // spec/protocol.md §5.9 requires table to match a §5.2 entry,
            // and list_tables' relkind = 'r' filter excludes partitioned
            // tables — so this must reject the same way an unknown table
            // name does, not silently answer [].
            expectBadRequest("?schema=$schema&table=events")
        } finally {
            dropSchema(schema)
        }
    }

    // finding #1: queryTable (and commonValues, same gate) validate table
    // via allowedTables, which used to match information_schema.tables'
    // broader BASE TABLE instead of listTables' relkind = 'r' — so a
    // partitioned table must be rejected here too, not silently queried.
    @Test
    fun `query table rejects a partitioned table same as referenced-by does`() {
        val schema = "ashb_test_spring_query_table_partitioning_gate"
        setupPartitionedSchema(schema)
        try {
            val ex = assertThrows(HttpClientErrorException::class.java) {
                http.getForObject(URI(url("/api/tables/data") + "?schema=$schema&table=events"), String::class.java)
            }
            assertEquals(HttpStatus.BAD_REQUEST, ex.statusCode)
        } finally {
            dropSchema(schema)
        }
    }
}

File: implementations/go-nethttp/referenced_by_integration_test.go
package ashurbanipal

import (
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"testing"
)

// GET /api/tables/referenced-by (spec/protocol.md §5.9) against the
// devcontainer's live Postgres with conformance/seed/seed.sql. Mirrors the
// Rust P5.9-* conformance checks (conformance/runner/referenced_by.rs):
// public.users is referenced by several tables incl. cross-schema
// warehouse.shipment_events; inventory_locations by exactly one composite
// FK; feature_flags by nothing. Presence-based for users, since its exact
// referrer set is large.

func referencedBy(t *testing.T, table string) []map[string]interface{} {
	t.Helper()
	body := getJSON(t, testServer(t), "/api/tables/referenced-by?table="+url.QueryEscape(table))
	raw, _ := body["referenced_by"].([]interface{})
	out := make([]map[string]interface{}, 0, len(raw))
	for _, e := range raw {
		out = append(out, e.(map[string]interface{}))
	}
	return out
}

func entriesFor(entries []map[string]interface{}, table string) []map[string]interface{} {
	var out []map[string]interface{}
	for _, e := range entries {
		if e["table"] == table {
			out = append(out, e)
		}
	}
	return out
}

// onlyPair asserts the entry carries exactly one {from, to} pair and returns it.
func onlyPair(t *testing.T, entry map[string]interface{}) (string, string) {
	t.Helper()
	cols, _ := entry["columns"].([]interface{})
	if len(cols) != 1 {
		t.Fatalf("expected exactly one column pair, got %v", entry["columns"])
	}
	c := cols[0].(map[string]interface{})
	return c["from"].(string), c["to"].(string)
}

func TestReferencedByListsIncomingFKConstraintsWithExplicitColumnPairs(t *testing.T) {
	entries := referencedBy(t, "users")

	orders := entriesFor(entries, "orders")
	if len(orders) != 1 {
		t.Fatalf("expected exactly one orders entry, got %d", len(orders))
	}
	if from, to := onlyPair(t, orders[0]); from != "user_id" || to != "id" {
		t.Errorf("orders pair = {%s, %s}, want {user_id, id}", from, to)
	}
	if s, _ := orders[0]["constraint"].(string); s == "" {
		t.Error("orders entry has an empty constraint")
	}
	if _, present := orders[0]["schema"]; present {
		t.Error("a same-schema referrer must omit schema")
	}

	// support_tickets has two separate FKs into users -> two entries.
	tickets := entriesFor(entries, "support_tickets")
	if len(tickets) != 2 {
		t.Fatalf("expected two support_tickets entries, got %d", len(tickets))
	}
	gotPairs := map[string]bool{}
	for _, e := range tickets {
		from, to := onlyPair(t, e)
		gotPairs[from+"->"+to] = true
		if _, present := e["schema"]; present {
			t.Error("support_tickets referrer must omit schema")
		}
	}
	if !gotPairs["user_id->id"] || !gotPairs["assigned_admin_id->id"] {
		t.Errorf("support_tickets pairs = %v, want user_id->id and assigned_admin_id->id", gotPairs)
	}

	// (table, constraint) is unique within one response.
	seen := map[string]bool{}
	for _, e := range entries {
		key := e["table"].(string) + "\x00" + e["constraint"].(string)
		if seen[key] {
			t.Errorf("duplicate (table, constraint): %s", key)
		}
		seen[key] = true
	}
}

func TestReferencedByIncludesCompositeForeignKeysWithEveryColumnPair(t *testing.T) {
	entries := referencedBy(t, "inventory_locations")
	if len(entries) != 1 {
		t.Fatalf("inventory_locations has exactly one referrer, got %d: %v", len(entries), entries)
	}
	e := entries[0]
	if e["table"] != "inventory_counts" {
		t.Errorf("referrer table = %v, want inventory_counts", e["table"])
	}
	cols, _ := e["columns"].([]interface{})
	var pairs [][2]string
	for _, c := range cols {
		m := c.(map[string]interface{})
		pairs = append(pairs, [2]string{m["from"].(string), m["to"].(string)})
	}
	want := [][2]string{{"warehouse_code", "warehouse_code"}, {"bin_code", "bin_code"}}
	if len(pairs) != len(want) {
		t.Fatalf("column pairs = %v, want %v", pairs, want)
	}
	for i := range want {
		if pairs[i] != want[i] {
			t.Errorf("column pair %d = %v, want %v (constraint order)", i, pairs[i], want[i])
		}
	}
}

func TestReferencedByReturnsEmptyListWhenNothingReferencesTheTable(t *testing.T) {
	if entries := referencedBy(t, "feature_flags"); len(entries) != 0 {
		t.Errorf("referenced_by(feature_flags) = %v, want []", entries)
	}
}

func TestReferencedByCarriesSchemaForCrossSchemaReferrers(t *testing.T) {
	se := entriesFor(referencedBy(t, "users"), "shipment_events")
	if len(se) != 1 {
		t.Fatalf("expected one shipment_events entry, got %d", len(se))
	}
	if se[0]["schema"] != "warehouse" {
		t.Errorf("shipment_events schema = %v, want warehouse", se[0]["schema"])
	}
	if from, to := onlyPair(t, se[0]); from != "handled_by_user_id" || to != "id" {
		t.Errorf("shipment_events pair = {%s, %s}, want {handled_by_user_id, id}", from, to)
	}
}

func TestReferencedByExplicitPublicSchemaMatchesImplicitDefault(t *testing.T) {
	implicit := getJSON(t, testServer(t), "/api/tables/referenced-by?table=users")
	explicit := getJSON(t, testServer(t), "/api/tables/referenced-by?schema=public&table=users")
	a, _ := json.Marshal(implicit["referenced_by"])
	b, _ := json.Marshal(explicit["referenced_by"])
	if string(a) != string(b) {
		t.Errorf("explicit schema=public differs from implicit:\n%s\n%s", a, b)
	}
}

func TestReferencedByRejectsUnknownOrMaliciousInputCleanly(t *testing.T) {
	base := testServer(t)
	bad := []string{
		"/api/tables/referenced-by?table=" + url.QueryEscape(""),
		"/api/tables/referenced-by?table=" + url.QueryEscape("no_such_table"),
		"/api/tables/referenced-by?table=" + url.QueryEscape(`users"; drop table users; --`),
		"/api/tables/referenced-by?table=" + url.QueryEscape("users' OR '1'='1"),
		"/api/tables/referenced-by", // table param missing
		"/api/tables/referenced-by?schema=no_such_schema&table=users",
		"/api/tables/referenced-by?source=no_such_source&table=users",
	}
	for _, path := range bad {
		resp, err := http.Get(base + "/__ashurbanipal" + path)
		if err != nil {
			t.Fatalf("GET %s: %v", path, err)
		}
		resp.Body.Close()
		if resp.StatusCode != http.StatusBadRequest {
			t.Errorf("GET %s: status %d, want 400", path, resp.StatusCode)
		}
		if got := resp.Header.Get("x-ashurbanipal-protocol"); got != "1" {
			t.Errorf("GET %s: protocol header = %q, want \"1\"", path, got)
		}
	}
}

func TestReferencedByCarriesProtocolHeaderOn200(t *testing.T) {
	resp, err := http.Get(testServer(t) + "/__ashurbanipal/api/tables/referenced-by?table=users")
	if err != nil {
		t.Fatalf("GET: %v", err)
	}
	defer resp.Body.Close()
	_, _ = io.Copy(io.Discard, resp.Body)
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("status %d, want 200", resp.StatusCode)
	}
	if got := resp.Header.Get("x-ashurbanipal-protocol"); got != "1" {
		t.Errorf("protocol header = %q, want \"1\"", got)
	}
}
